core/circuit.go

```diff
@@ -0,0 +1,262 @@
+package core
+
+import (
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+)
+
+var gridMeterUsed bool // indicates gridmeter is used already fo a circuit to avoid > 1 usage
+
+type Circuit struct {
+	Log    *util.Logger
+	uiChan chan<- util.Param
+
+	Name       string    `mapstructure:"name"`       // meaningful name
+	MaxCurrent float64   `mapstructure:"maxCurrent"` // the max allowed current of this circuit
+	MeterRef   string    `mapstructure:"meter"`      // Charge meter reference
+	Circuits   []Circuit `mapstructure:"circuits"`   // sub circuits as config reference
+
+	parentCircuit *Circuit         // parent circuit reference
+	meterCurrent  api.MeterCurrent // meter to determine phase current
+	vMeter        *VMeter          // virtual meter if no real meter is used
+}
+
+// determines current in use.
+// implement consumer interface
+// TBD: phase perfect evaluation
+func (cc *Circuit) GetCurrent() (float64, error) {
+	if cc.meterCurrent == nil {
+		cc.Log.ERROR.Println("meterCurrent is nil")
+		return 0.0, fmt.Errorf("no meter available")
+	}
+
+	var current float64
+	var err error
+	i1, i2, i3, err := cc.meterCurrent.Currents()
+	if err == nil {
+		cc.Log.DEBUG.Printf("meter currents: %.3gA", []float64{i1, i2, i3})
+		cc.publish("meterCurrents", []float64{i1, i2, i3})
+		// TODO: phase adjusted handling. Currently we take highest current from all phases
+		current = math.Max(math.Max(i1, i2), i3)
+	} else {
+		cc.Log.ERROR.Println(fmt.Errorf("grid meter currents: %v", err))
+		return 0.0, fmt.Errorf("error getting meter currents: %v", err)
+	}
+
+	cc.Log.DEBUG.Printf("actual current: %.1fA", current)
+	cc.publish("actualCurrent", current)
+	return current, nil
+}
+
+// returns avaialble current based on limit and consumption
+// todo: GetRemainingCurrent
+func (cc *Circuit) GetRemainingCurrent() float64 {
+	cc.Log.TRACE.Printf("get available current")
+	// first update current current, mainly to regularly publish the value
+	current, _ := cc.GetCurrent()
+	curAvailable := cc.MaxCurrent - current
+	if curAvailable < 0.0 {
+		cc.Log.WARN.Printf("overload detected (%s) - currents: %.1fA, allowed max current is: %.1fA\n", cc.Name, current, cc.MaxCurrent)
+		cc.publish("overload", true)
+	} else {
+		cc.publish("overload", false)
+	}
+	// check parent circuit, return lowest
+	if cc.parentCircuit != nil {
+		cc.Log.TRACE.Printf("get available current from parent: %s", cc.parentCircuit.Name)
+		curAvailable = math.Min(curAvailable, cc.parentCircuit.GetRemainingCurrent())
+	}
+	cc.Log.DEBUG.Printf("circuit using %.1fA, %.1fA available", current, curAvailable)
+	return curAvailable
+}
+
+// creates a circuit with defaults
+func NewCircuit(n string, limit float64, mc api.MeterCurrent, l *util.Logger) *Circuit {
+	cc := &Circuit{
+		Name:         n,
```

Comment:
Warum brauchts hier den Namen? Sonst kennt kein Objekte seinen Namen

Reply:
im Log siehts besser aus wenn da der Name des Circuit steht wie in der Config vom user angegeben als `cc-0`. Fände ich auf für die Loadpoints besser ...